### DISCUSSION TITLE
Handeling UserAgent / DOM errors

### DISCUSSION THREAD

**Discussion by hussam-qasem** [mojolicious/mojo#2063] [Q&A] [answered]
1. How do I detect `UserAgent` errors? (I can't seem to catch them). For example:
```perl
my $ua = Mojo::UserAgent->new;
$ua->proxy->https('http://127.0.0.1:3128');
my $tx = $ua->get( "https://www.mojolicious.org/index.html" );
die $tx->result->message unless $tx->result->is_success;
```
produces `Proxy connection failed`. How can I capture this error?

2. What's the preferred way to handle `DOM` errors? For example:
```perl
eval {
    say STDOUT "name: ",
	Mojo::DOM->new
	->parse( '<tr><td>Username:</td><td><input type="text" Xname="g3x761111" /></td></tr>' )
	->at( 'input[type="text"][name^="g3"]' )->attr->{name};
};

if ($@) {
    warn "Error accessing input element: $@";
}
```
vs.
```perl
my $dom = Mojo::DOM->new
  ->parse( '<tr><td>Username:</td><td><input type="text" Xname="g3x761111" /></td></tr>' )
  ->at( 'input[type="text"][name^="g3"]' );
say STDOUT "name: ", $dom->attr->{name} if $dom;
warn "Error accessing input element" unless $dom;
```

**Comment by hussam-qasem** [comment]
>     1. How do I detect `UserAgent` errors? (I can't seem to catch them). For example:
> 
> 
> ```perl
> my $ua = Mojo::UserAgent->new;
> $ua->proxy->https('http://127.0.0.1:3128');
> my $tx = $ua->get( "https://www.mojolicious.org/index.html" );
> die $tx->result->message unless $tx->result->is_success;
> ```
> 
> produces `Proxy connection failed`. How can I capture this error?

So I figured one part. It seems accessing the `result` object throws the exception.
```perl
my $ua = Mojo::UserAgent->new;
$ua->proxy->https('http://127.0.0.1:3128');
my $tx = $ua->get( "https://www.mojolicious.org/index.html" );
my $result = eval { $tx->result };
die "Error occurred: $@" if $@;
die $result->code."/".$result->message unless $result->is_success;
```

**Reply by daleif** [reply]
Erh, did you have a look at https://docs.mojolicious.org/Mojo/UserAgent, wouldn't the other messages be able to catch or detect these?

**Reply by hussam-qasem** [reply]
Thanks for your response. Would you kindly elaborate with an example? I don't see anything relevant in the [`UserAgent`](https://docs.mojolicious.org/Mojo/UserAgent) documentation. However, [`Mojo::Exception`](https://docs.mojolicious.org/Mojo/Exception) provides some insight:
```perl
#!/usr/bin/env perl
use Mojo::UserAgent;
use Mojo::Exception qw(check);

my $ua = Mojo::UserAgent->new;
$ua->proxy->https('http://127.0.0.1:3128');
my $tx = $ua->get( "https://www.mojolicious.org/index.html" );;
eval { $tx->result };
check $@ => [
    default => sub { die "Error occurred: $_" },
    ];
die $tx->result->code."/".$tx->result->message unless $tx->result->is_success;
```
I'm using `default` because I don't know how to get the actual exception name.

**Reply by kraih** [reply, marked as answer]
The whole point `->result` exists is to throw an exception. If you don't want that you could just call `->res` instead and check manually for errors.

**Reply by hussam-qasem** [reply]
Thank you @kraih for the prompt response. I was not aware of the difference - I now see it in [`Mojo::Transaction`](https://docs.mojolicious.org/Mojo/Transaction)
